### ISSUE TITLE
Tracking issue for UnixFS automatic sharding

### ISSUE THREAD

**Issue by aschmahmann** [ipfs/go-ipfs#8106]
**Update 10/21**: All major work has been finished. Full review pending.

The sharding PR has landed and we're working on the reverse transition of "unsharding". We haven't merged it in go-ipfs (https://github.com/ipfs/go-ipfs/pull/8114) until we finish with the unsharding to have the whole functionality together (and ensure deterministic CIDs for any directory size).

Main unsharding work happening in https://github.com/ipfs/go-unixfs/pull/94.

Minor issues left for last compiled in https://github.com/ipfs/go-unixfs/issues/105 can be addressed after main PRs land.

Potential issue with `merkledag.Walk` function doing a BFS in https://github.com/ipfs/go-unixfs/issues/107, but this can be address after landing the major work as it's only an optimization.

-----------------------------------------------

Tracking implementation of https://github.com/ipfs/go-ipfs/issues/7022

Per the linked issue in order to enable sharded directories by default we want to shard the directories when required.

After some discussion the heuristic we are going to use for now to do determine when to use sharded vs regular directories is option three listed [here](https://github.com/ipfs/go-ipfs/issues/7022#issuecomment-831490209). That is we will sum the sizes of all the names + CIDs in the map and if they are >256KiB we will use sharded directories and if they are <=256KiB we will use regular directories.

Places where we'll likely have to look at changes include:
- go-unixfs (the importer and directory code as well as potentially the helper code)
- go-mfs (wherever we modify a folder)

Hopefully if we keep the interfaces the same here we won't have to make any changes in go-ipfs itself. We may need to do a scan to see if we do any type casting to a UnixFS basic or sharded directory though just in case.

Once this is done we should be able to drop the global boolean that enables "use sharded directories" from both go-unixfs, go-ipfs, and go-ipfs-config.

**Comment by schomatis** [comment]
Scoped issue with technical details in https://github.com/ipfs/go-mfs/issues/87 regarding the MFS/UnixFS enhancement to support this. Once that is done `go-ipfs` should only need to update the dependencies and set the new option with the desired 256KiB value.

**Comment by schomatis** [comment]
We still need to integrate this into `go-ipfs` (draft in https://github.com/ipfs/go-ipfs/pull/8114), namely fixing sharness and interop tests.

**Comment by BigLep** [comment]
2021-05-10 discussion: 
1. @schomatis is going to do the unsharding work week of 2021-05-17
2. @mburns is going to do a test fix here.  @schomatis or @aschmahmann can provide the details.

**Comment by schomatis** [comment]
More details in the TODO list of the ongoing PR https://github.com/ipfs/go-ipfs/pull/8114.

**Comment by schomatis** [comment]
Update 8/13 (@schomatis): I'm leading this effort currently in progress. _Not blocked_.

* The sharding PR has landed and we're working on the reverse transition of "unsharding". We haven't merged that in go-ipfs (https://github.com/ipfs/go-ipfs/pull/8114) until we finish with the unsharding to have the whole functionality together (and ensure deterministic CIDs for any directory size).

* Main unsharding work happening in https://github.com/ipfs/go-unixfs/pull/94. There are several reviews that I need to address next week.The PR is constantly being updated with the main TODO items.

* The tests have been progressing in a separate PR in https://github.com/ipfs/go-unixfs/pull/99 which now have a preliminary code for review. After https://github.com/ipfs/go-unixfs/pull/94 is lands will re-focus here.

**Comment by schomatis** [comment]
**Update 8/17** (@schomatis, DRI):

Brief:
* **Not blocked.**
* Reviews pending for:
  * https://github.com/ipfs/go-unixfs/pull/102
  * https://github.com/ipfs/go-unixfs/pull/103
  * https://github.com/ipfs/go-unixfs/pull/99

**Comment by schomatis** [comment]
**Update 8/17** (corrected) (@schomatis, DRI):

Brief:
* **Not blocked.** Working on https://github.com/ipfs/go-unixfs/pull/94.
* Reviews pending for:
  * https://github.com/ipfs/go-unixfs/pull/102
  * https://github.com/ipfs/go-unixfs/pull/103
  * https://github.com/ipfs/go-unixfs/pull/99

**Comment by schomatis** [comment]
**Update 8/23** (@schomatis, DRI):

Brief:
* Working on tests in separate [branch](https://github.com/ipfs/go-unixfs/tree/origin/schomatis/directory/unsharding-tests-extra) while https://github.com/ipfs/go-unixfs/pull/94 gets reviewed.
* Reviews pending for:
  * https://github.com/ipfs/go-unixfs/pull/102
  * https://github.com/ipfs/go-unixfs/pull/103
  * https://github.com/ipfs/go-unixfs/pull/99

(See full status in the OP.)

**Comment by schomatis** [comment]
**Update 9/3** (@schomatis, DRI):

Brief:
* Working on concurrent test in https://github.com/ipfs/go-unixfs/pull/106.
* Reviews pending for:
  * https://github.com/ipfs/go-unixfs/pull/106
  * Potential issue with `merkledag.Walk` function doing a BFS: https://github.com/ipfs/go-unixfs/issues/107.
  * https://github.com/ipfs/go-unixfs/pull/99 (Need a more detailed review on `io/directory_test.go` as explained in PR description.)

(See full status in PR description.)

**Comment by aschmahmann** [comment]
Closed by #8563
